spec/data-model.md

```diff
@@ -0,0 +1,183 @@
+
+To work with messages defined in other syntaxes than that of MessageFormat 2,
+an equivalent data model representation is also defined.
+Implementations MAY provide interfaces which allow
+for MessageFormat 2 syntax to be parsed into this representation,
+for this representation to be serialized into MessageFormat 2 syntax
+or any other syntax,
+for messages presented in this representation to be formatted,
+or for other operations to be performed on or with messages in this representation.
+
+Implementations are not required to use this data model for their internal representation of messages.
+
+To ensure compatibility across all platforms,
+this interchange data model is defined in terms of JSON-compatible values.
+While this document uses TypeScript syntax for their definition,
+the canonical and authoritative source is the `message.json`
+[JSON Schema](https://json-schema.org/) definition.
+
+## Messages
+
+A `SelectMessage` corresponds to a syntax message that includes _selectors_.
+A message without _selectors_ and with a single _pattern_ is represented by a `PatternMessage`.
+
+```ts
+type Message = PatternMessage | SelectMessage
+
+interface PatternMessage {
+  type: 'message'
+  declarations: Declaration[]
+  pattern: Pattern
+}
+
+interface SelectMessage {
+  type: 'select'
+  declarations: Declaration[]
+  selectors: Expression[]
+  variants: Variant[]
+}
+```
+
+Each message _declaration_ is represented by a `Declaration`,
+which connects the `name` of the left-hand side _variable_
+with its right-hand side `value`.
+The `name` does not include the initial `$` of the _variable_.
+
+```ts
+interface Declaration {
+  name: string
+  value: Expression
+}
+```
+
+In a `SelectMessage`,
+the `keys` and `value` of each _variant_ are represented as an array of `Variant`.
+For the `CatchallKey`, a string `value` may be provided to retain an identifier.
+This is always `'*'` in MessageFormat 2 syntax, but may vary in other formats.
+
+```ts
+interface Variant {
+  keys: Array<Literal | CatchallKey>
+  value: Pattern
+}
+
+interface CatchallKey {
+  type: '*'
+  value?: string
+}
+```
+
+## Patterns
+
+Each `Pattern` represents a linear sequence, without selectors.
+Each element of the sequence MUST have either a `Text` or an `Expression` shape.
+`Text` represents literal _text_,
+while `Expression` wraps each of the potential _expression_ shapes.
+The `value` of `Text` is the "cooked" value (i.e. escape sequences are processed).
+
+Implementations MUST NOT rely on the set of `Expression` `body` values being exhaustive,
+as future versions of this specification MAY define additional expressions.
+If encountering a `body` with an unrecognised value,
+an implementation SHOULD treat it as it would a `Reserved` value.
+
+```ts
+interface Pattern {
+  body: Array<Text | Expression>
+}
+
+interface Text {
+  type: 'text'
+  value: string
+}
+
+interface Expression {
+  type: 'expression'
+  body: Literal | VariableRef | FunctionRef | Reserved
```

Comment:
A `FunctionRef` can also have a `Literal` or a `VariableRef` as an argument, and in fact, due to how our syntax is designed, I'd argue that the function's argument is more important than the function. (E.g. it comes first in the syntax.)

I'd like to suggest an alternative way to structure our _expressions_, to more closely map to our syntax:

```abnf
expression = "{" [s] ((operand [s annotation]) / annotation) [s] "}"
```

Let's special-case argument-less functions rather than function-less operands.

Instead of `Literal | VariableRef | FunctionRef` here and `operand?: Literal | VariableRef` inside `FunctionRef`, we can do:

```ts
type Expression = OperandExpr | FunctionExpr;
interface OperandExpr {
    operand: Literal | VariableRef;
    annotation?: FunctionExpr;
}
interface FunctionExpr {
    operand: Literal | VariableRef;
    annotation?: FunctionExpr;
}
```

FWIW, this is how I implemented expressions in [stasm/message2](https://github.com/stasm/message2): 
https://github.com/stasm/message2/blob/4abf43f2023b6e20d8ee1d462684d0741ece791b/syntax/ast.ts#L44-L70

(Not blocking this PR on this, but I'd like to discuss this change as a follow-up.)

Reply:
I will be happy to discuss this further in a follow-on issue or PR.

Reply:
Filed #426 to continue this.